synapse/federation/transport/server/federation.py

```diff
@@ -113,7 +113,7 @@ async def on_PUT(
                 device_list_updates = [
                     edu.content
                     for edu in transaction_data.get("edus", [])
-                    if edu.edu_type in DEVICE_UPDATE_EDUS
+                    if edu.get("edu_type") in DEVICE_UPDATE_EDUS
```

Comment:
This could still error if the `edu_type` is weird (e.g. a list). A way around this is making `DEVICE_UPDATE_EDUS` into a `list`, not a `set`.

Reply:
Oh I see, I'm implicitly assuming that the edu_type is hashable.

Reply:
Yep! It is *probably* OK, but if the downside is breaking federation...that's pretty bad.

Reply:
That's done. I haven't fixed up the TX side here

https://github.com/matrix-org/synapse/blob/f160fe18e3cc5604375f491300d12dd5c7e9b9b2/synapse/federation/sender/transaction_manager.py#L128-L138

because there we really do have an `Edu` struct, and its type is something we should be fully in control of.